src/ucp/core/ucp_mm.c

```diff
@@ -839,40 +888,50 @@ ucs_status_t ucp_mem_map(ucp_context_h context, const ucp_mem_map_params_t *para
                                 uct_flags, "user memory", &memh);
     } else {
         status = ucp_memh_create(context, address, length, mem_type,
-                             UCT_ALLOC_METHOD_LAST, 0, &memh);
+                                 UCT_ALLOC_METHOD_LAST, 0, &memh);
         if (status != UCS_OK) {
             goto out;
         }
 
         status = ucp_memh_init_uct_reg(context, memh, uct_flags);
         if (status != UCS_OK) {
             ucs_free(memh);
-            goto out;
         }
     }
 
+out:
     if (status == UCS_OK) {
         ucs_assert(memh->md_map != 0);
-        ucs_assert(memh->parent != NULL);
+        ucs_assert(ucp_memh_is_user_memh(memh));
+        *memh_p = memh;
     }
-
-out:
-    *memh_p = memh;
     return status;
 
 out_zero_mem:
     ucs_assert(ucp_memh_address(&ucp_mem_dummy_handle.memh) == NULL);
     ucs_assert(ucp_memh_length(&ucp_mem_dummy_handle.memh) == 0);
 
     ucs_debug("mapping zero length buffer, return dummy memh");
-    memh   = &ucp_mem_dummy_handle.memh;
-    status = UCS_OK;
-    goto out;
+    *memh_p = &ucp_mem_dummy_handle.memh;
+    return UCS_OK;
+}
+
+void ucp_memh_put_slow(ucp_context_h context, ucp_mem_h memh)
+{
+    ucs_assert(context->rcache == NULL);
+    ucs_assert(memh->parent == NULL);
+    ucp_memh_dereg(context, memh, memh->md_map);
+    ucs_free(memh);
 }
 
 ucs_status_t ucp_mem_unmap(ucp_context_h context, ucp_mem_h memh)
 {
-    ucp_memh_put(context, memh);
+    if (ucp_memh_is_zero_length(memh)) {
+        return UCS_OK;
+    }
+
+    ucp_memh_cleanup(context, memh);
+    ucs_free(memh);
```

Comment:
maybe move ucs_free inside ucp_memh_cleanup, and then call ucp_memh_cleanup from mpool release callbacks (instead of ucp_mem_unmap)
moreover, IMO we could unite ucp_memh_cleanup and ucp_memh_cleanup to one function to handle both cases (or call ucp_memh_put_slow from ucp_memh_cleanup)